Fix transitive unavailability check

tjw opened the pull request:
Allow calls to explicitly unavailable code when the caller has the
same unavailability.

For example, in a framework compiled with `-application-extension`, this should
be allowed:

	@available(iOSApplicationExtension, unavailable)
	    func x() {
            _ = UIApplication.shared
	    }
	}

since the consumer of the framework, or other call sites within the same
framework will be unable to call the unsafe wrapper in the same situations they
couldn't call the original unsafe code.

Resolves [SR-1226](https://bugs.swift.org/browse/SR-1226).

tjw committed:
Possible fix for SR-1226.

Allow calls to explicitly unavailable code when the caller has the
same unavailability.

For example, in a framework compiled with `-application-extension`, this should
be allowed:

	@available(iOSApplicationExtension, unavailable)
	    func x() {
            _ = UIApplication.shared
	    }
	}

since the consumer of the framework, or other call sites within the same
framework will be unable to call the unsafe wrapper in the same situations they
couldn't call the original unsafe code.

tjw commented:
Note that there is a stricter early-out above the new check that should maybe be removed:

    // Suppress the diagnostic if we are in synthesized code inside
    // a synthesized function and the reference is lexically
    // contained in a declaration that is itself marked unavailable.
    // The right thing to do here is to not synthesize that code in the
    // first place. rdar://problem/20491640
    if (R.isInvalid() && isInsideImplicitFunction(R, DC) &&
        isInsideUnavailableDeclaration(R, DC)) {
      return false;
    }

jrose-apple commented on lib/Sema/TypeCheckAvailability.cpp:
```diff
@@ -2093,6 +2093,14 @@ bool TypeChecker::diagnoseExplicitUnavailability(
+  // unavailability is OK -- the eventual caller can't call the
+  // enclosing code in the same situations it wouldn't be able to
+  // call this code.
+  if (isInsideUnavailableDeclaration(R, DC)) {
```
I think this is still a problem: it would let something guarded with `@available(iOSApplicationExtension, unavailable)` call things that were `@available(iOS, unavailable)`. Modeling application extension restrictions as availability really needs more attention in Swift.

tjw committed:
Only skip unavailability diagnostics if the platforms match.

Add TypeChecker::isInsideCompatibleUnavailableDeclaration().
Pass in the call site's AvailableAttr and compare its platform to those
in the enclosing lexical scopes.

Added some basic tests.

gottesmm commented on lib/Sema/TypeCheckAvailability.cpp:
```diff
@@ -1499,6 +1499,27 @@ bool TypeChecker::isInsideUnavailableDeclaration(
+bool TypeChecker::isInsideCompatibleUnavailableDeclaration(
+	SourceRange ReferenceRange, const DeclContext *ReferenceDC,
+	const AvailableAttr *attr)
+{
```
Can you be sure to run this through clang-format? This is not LLVM style. If you use git-clang-format, everything will just work for you.

tjw replied:
I don't see any mention of formatting in [Contributing Code](https://swift.org/contributing/#contributing-code). Is there another page of documentation that covers this (and possibly other conventions I've missed)?

tjw replied:
Ah, or maybe Swift just adopts all the conventions from clang... =)

jrose-apple replied:
That's it, or at least 90% of them, but we should definitely mention that on the page.

gottesmm reviewed (commented):
Quick drive by comment inline and more generally be sure to rebase your changes before merging.

jrose-apple commented on lib/Sema/TypeCheckAvailability.cpp:
```diff
@@ -1499,6 +1499,27 @@ bool TypeChecker::isInsideUnavailableDeclaration(
+	  return EnclosingUnavailable && EnclosingUnavailable->Platform == platform;
+    };
+
+    return someEnclosingDeclMatches(ReferenceRange, ReferenceDC, *this,
```
Hm, this doesn't quite seem like the right query. If a declaration is unavailable on every platform, and then the *next* enclosing declaration is unavailable in a compatible way, this will return `true` rather than `false`.

tjw replied:
I'm not sure I understand, but here goes ... my understanding is that the `Platform` of the `AvailableAttr` is always either `::none` or the *current compiling* platform by this point. It seems like `AvailableAttr` holds the sum of the reasons the current declaration is unavailable in the current compile (introduction/deprecation versions, platform), but it does not hold a complete record of every `@available` annotation in the source.

So, when I try to add a test that sounds like what you are suggesting, it works as I would hope:

	// RUN: %target-typecheck-verify-swift
	// REQUIRES: OS=macosx

	// Check that only the platform we are compiling for is considered when allowing access to unavailable declarations.

	@available(iOS, unavailable)
	@available(OSX, unavailable)
	func unavailable() {} // expected-note {{'unavailable()' has been explicitly marked unavailable here}}

	@available(iOS, unavailable)
	func call_with_ios_unavailable() {
		unavailable() // expected-error {{'unavailable()' is unavailable}}
	}

	@available(OSX, unavailable)
	func call_with_osx_unavailable() {
		unavailable() // OK: same
	}

Adding debugging code to `TypeChecker::isInsideCompatibleUnavailableDeclaration()`, it is called twice. In both cases the `Platform` is 4 (`OSX`). That is, the `@available(iOS, unavailable)` enclosing the first call site has been discarded by this point since we aren't compiling for iOS, and it doesn't permit the call to `unavailable()` from `call_with_ios_unavailable()`.

jrose-apple replied:
Sorry, I should have just included a code example.

```swift
// on macOS
@available(iOS, unavailable)
class Outer {
  @available(*, unavailable)
  func completelyBadMethod() {}
}

@available(iOS, unavailable)
func test(outer: Outer) {
  outer.completelyBadMethod() // should still error
}
```

On top of that, your example is also unfortunate, because it means the active and non-active platforms don't behave the same. But I think that's inevitable, since you can always imagine a hypothetical dishwasherOS platform that can call everything.

tjw replied:
Ah, got it. Test added and passes locally. Also added a swapped version of the nested availability, which fails to build as expected since the outer declaration can't be referenced.

jrose-apple replied:
Ah, hm. Now I need to figure out why this works. I really thought it would identify the outer context as being compatible and then return true.

tjw replied:
I believe it is that the 'completely unavailable' declaration passes the first check:

	if (!attr->isUnconditionallyUnavailable()) {
		return false;
	}

but fails the second:

	PlatformKind platform = attr->Platform;
	if (platform == PlatformKind::none) {
		return false;
	}

and so we don't look at any enclosing source ranges.

tjw replied:
It seems possible that a more complicated stack of nested availabilities might provoke the issue you were worried about. I'll can poke at it a bit more after my current `--test` build finishes.

tjw replied:
So far I haven't been able to write a test that breaks. Any stack of iOS/unavailable/iOS stops at the point the caller tries to spell the unavailable symbols and doesn't look at the child symbols. And the test you suggested hits the path mentioned above.

Perhaps I should add some commentary to make the intention more clear.

tjw replied:
Not sure sure if this is more clear for future readers or not... thoughts?


	// If a declaration has an unavailable attribute applied, it can still
	// called if some enclosing scope of the call site has the same unavailability.
	bool TypeChecker::isInsideCompatibleUnavailableDeclaration(
	    SourceRange ReferenceRange, const DeclContext *ReferenceDC,
	    const AvailableAttr *attr) {
	  // We aren't attempting to do introduced/deprecated version checks
	  if (!attr->isUnconditionallyUnavailable()) {
	    return false;
	  }
  
	  // Only allow per-platform unavailability for the current platform.
	  // At this point, the platform will either be ::none (totally unavailable)
	  // and not allowed here, or the current platform. All other annotations
	  // of `@unavailable(OtherPlatform, *)` will have been dropped already.
	  PlatformKind platform = attr->Platform;
	  if (platform == PlatformKind::none) {
	    return false;
	  }

	  auto IsUnavailable = [platform](const Decl *D) {
	    auto EnclosingUnavailable =
	        D->getAttrs().getUnavailable(D->getASTContext());
	    return EnclosingUnavailable && EnclosingUnavailable->Platform == platform;
	  };

	  return someEnclosingDeclMatches(ReferenceRange, ReferenceDC, *this,
	                                  IsUnavailable);
	}

jrose-apple reviewed (commented):
This seems like a good approach! I don't think it's quite there yet, though, which also makes me suggest adding more test cases with nesting.

gottesmm commented:
> On Oct 20, 2017, at 4:22 PM, Timothy J. Wood <notifications@github.com> wrote:
> 
> @tjw commented on this pull request.
> 
> In lib/Sema/TypeCheckAvailability.cpp <https://github.com/apple/swift/pull/12410#discussion_r146085298>:
> 
> > @@ -1499,6 +1499,27 @@ bool TypeChecker::isInsideUnavailableDeclaration(
>                                    IsUnavailable);
>  }
>  
> +bool TypeChecker::isInsideCompatibleUnavailableDeclaration(
> +	SourceRange ReferenceRange, const DeclContext *ReferenceDC,
> +	const AvailableAttr *attr)
> +{
> I don't see any mention of formatting in Contributing Code <https://swift.org/contributing/#contributing-code>. Is there another page of documentation that covers this (and possibly other conventions I've missed)?
> 
> 

We should add it. Generally people just fire and forget with git-clang-format.
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub <https://github.com/apple/swift/pull/12410#discussion_r146085298>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AAee39UV7d9vnQpDoDX8aJQqXpTZwf6rks5suSsogaJpZM4P3vBw>.
>

tjw committed:
Fix formatting.

tjw committed:
Test @available annotations from multiple platforms.

tjw committed:
Add nested availability check.

This helps ensure that a matching unavailability doesn't incorrectly allow
a reference to a completely unavailable declaration.

tjw commented:
Ping... is there anything I can do to make this patch better? Thoughts on the possible comment above ... not sure if it is clarifying or not.

tjw commented:
Building a local toolchain and testing on more of our code exposes a case I'm not handling yet:

    @available(iOSApplicationExtension, unavailable)
    let property: String = "qqq"

    @available(iOSApplicationExtension, unavailable)
    public let strings_let: [String] = [property]

jrose-apple commented:
Sorry, it's been a busy week. I'll take another look tomorrow and see if I can find the elusive case I cared about, or prove to myself that it's not possible.

When you say "a case you're not handling yet", that means this is rejected where it ought to be allowed, right? That's okay since it's not a regression; we can get it in a follow-up patch.

tjw commented:
Correct -- it is a case that should be allowed, but is currently (and was previously) rejected.

jrose-apple commented:
Okay, I think I've convinced myself that in order to access something more available inside something less available, you have to either name the enclosing type, or that type has to be named somewhere that's already available. Which means I'm good with this.

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
@swift-ci Please test source compatibility

jrose-apple commented:
Merged. Thank you, Timothy!
